### PR TITLE
m(windows): Reimplement Wepoll in Rust

### LINKED ISSUE #22: Removing the wepoll dependency?
Would it be possible to implement the windows backend directly in rust? I think making the code as accessible as possible to users is especially important with a library like this (mostly for debugging).

It looks like `mio` is written with `{nt, win}api`, and maybe it could be used as reference here.

### LINKED ISSUE #85: Null pointer dereference in patched wepoll code
See piscisaureus/wepoll#32. It looks like the patched wepoll code has a null pointer dereference for when the poller is notified.

### PR DESCRIPTION
**Pull request by notgull**
Thus far, `wepoll` has served us well in terms of polling on Windows. However, at this stage, it's time to migrate to our own home-rolled implementation. There is at least one critical soundness bug that stems from our usage of `wepoll` (#85), and it would be significantly easier to compile `polling` (as well as all of its dependents) if we purged our only C dependency. In addition, there are some issues that we can only solve if we move away from `wepoll` (smol-rs/async-io#17, smol-rs/async-io#25, #82).

The new strategy is similar to what `wepoll` currently does, except (hopefully) more robust. The core consists of using an I/O Completion Port as the core of the system, and then using the undocumented Auxillary Function Driver to deliver completion packets whenever a socket is considered "ready". There is a state object associated with each socket that contains the events currently being polled, the in-flight events associated with the socket, and whether the socket state is queued for deletion.

The differences between this implementation and `wepoll` includes:

- Rather than using a red-black tree to store socket state, we use a `HashMap`.
- A bounded `ConcurrentQueue` is used to hold pending socket updates, rather than a naive queue implementation. This queue has a capacity of 1024 elements and every update is run run the queue is full. In contrast, `wepoll`'s queue is unbounded.
- Packets have `Pin` guarantees, in order to prevent state captured by the I/O completion port from being moved.
- Instead of going through the `epoll_event` intermediary, we directly convert completion packets to `Event`s.
- We directly support notifying the `Poller` through a designated wakeup packet.
- We only expect one user to be polling at a time, so we use an `AtomicBool` instead of an atomic number.

Closes #22 and closes #85. Supersedes #76.

### CODE REVIEW

**Comment by notgull** [comment]
I think this is stable enough for review now; the tests aren't catching any bugs that I can see (although the `precision` tests are taking a while? But that might just be part of the test).

**Review by taiki-e — COMMENTED** [review]
Thanks! I have not reviewed the details regarding the Windows API, but at first glance, this looks great.

Also, we may want to test some downstream crates before merging/releasing this patch because such a rewrite may cause subtle differences in behavior. (in the similar [way that our CI tests async-io](https://github.com/smol-rs/polling/blob/d443196f641dfe83904dcc8bc8e163ec64180cdd/.github/workflows/ci.yml#L52-L59))

File: src/iocp/afd.rs
Comment (notgull) [reply src/iocp/afd.rs:563]: My bad, that was a poorly written safety comment. It's more like "the socket that's returned by the IOCTL must be valid." I changed it to be more clear.

**Comment by notgull** [comment]
> Also, we may want to test some downstream crates before merging/releasing this patch because such a rewrite may cause subtle differences in behavior.

I just ran a small "crater test" on a few crates that have `polling` in their dependency tree. I ran `cargo test` for:

- `async-io`
- `async-net`
- `async-std`
- `isahc` 
- `libp2p` (had to figure out how to set up `protoc` on Windows for this 😮‍💨)
- `quinn` (tests mostly use `tokio` but the ones I modified to use `async-std` seemed to turn out OK)
- `zbus`

All of the tests seem to pass (or at least the ones that weren't already failing on Windows). I think we're fine in terms of potentially undefined behavior.

File: src/iocp/port.rs
Comment (fogti) [inline src/iocp/port.rs:275]: what are the safety requirements of this function?

File: src/iocp/port.rs
Comment (fogti) [inline src/iocp/port.rs:84]: why is the pointer that this function returns pinned?

File: src/iocp/afd.rs
Comment (fogti) [inline src/iocp/afd.rs:450]: why do we roll our own `OnceCell` implementation here?

File: src/iocp/afd.rs
Comment (fogti) [inline src/iocp/afd.rs:603]: is this really the only possible error?

File: src/iocp/afd.rs
Comment (fogti) [inline src/iocp/afd.rs:589]: what do we gain by making this `u32` `MaybeUninit` here?

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:265]: this won't be run during release builds, right?

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:324]: so, will this then behave correctly on release builds, when `self.polling` never gets modified?

File: src/iocp/port.rs
Comment (notgull) [reply src/iocp/port.rs:285]: It should only be called once, since it moved out the `T`, like `ManuallyDrop::take()`. I've clarified this in the docs.

File: src/iocp/port.rs
Comment (notgull) [reply src/iocp/port.rs:88]: The pinning ensures that the overlapped object isn't moved while it's involved in an operation.

File: src/iocp/afd.rs
Comment (notgull) [reply src/iocp/afd.rs:450]: We can't use `once_cell` since its MSRV is 1.56, and pulling in `async_lock::OnceCell` would add a few too many dependencies for my liking. I figured rolling my own is the best option.

File: src/iocp/afd.rs
Comment (notgull) [reply src/iocp/afd.rs:603]: Yes; it's basically like returning `-1` for C functions.

File: src/iocp/afd.rs
Comment (notgull) [reply src/iocp/afd.rs]: Nothing, I guess. Fixed.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs]: I'm not sure, I *think* the values are still consumed even on release mode? Either way, it's better to make it more explicit. Fixed!

File: src/iocp/port.rs
Comment (fogti) [reply src/iocp/port.rs:88]: how could it be moved when it is behind an `&` reference?

File: src/iocp/afd.rs
Comment (fogti) [reply src/iocp/afd.rs:563]: huh, did the safety comment vanish?

File: src/iocp/port.rs
Comment (notgull) [reply src/iocp/port.rs:88]: While it's being used in the overlapped operation, there is no reference (or at least, none that the borrow checker is aware of). As a safeguard against undefined behavior, this ensures that it isn't invalidated during the operation.

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:269]: same here.

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:417]: where is this used and why does it always clone packets?

File: src/iocp/afd.rs
Comment (notgull) [reply src/iocp/afd.rs:563]: It's here:

https://github.com/smol-rs/polling/blob/c3ab16109ce2a86aaafe1717c56deb3c83a5f9de/src/iocp/afd.rs#L586

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:616]: this nesting is a bit hard to read, spreading it out over multiple lines (including the currently implicit `drop`) might be clearer.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs:417]: It's used in `lib.rs` for returning packets:

https://github.com/smol-rs/polling/blob/e85331c437eb5b9fce382e6a9a310e44a641a3d6/src/lib.rs#L500

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:638]:
```suggestion
    event.readable &= socket_state.interest.readable;
    event.writable &= socket_state.interest.writable;
```
would imo be easier to comprehend

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:668]: this message could be confusing, as it is missing context that is also not obvious from the function name. A clear indication that the notification packet couldn't be deleted because it wasn't a socket might be a good idea.

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:171]: instead of doing such a weird dance around poisoning all the time in this code, wouldn't it be more appropriate to either pull in a library that already implements locking in such a way (e.g. `parking-lot`) or write an abstraction around `RwLock` (and maybe `Mutex`) once, that does this, and use it in all this code?

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:219]: this might be cleaner done via chaining, e.g.
```suggestion
            sources.get(&socket)
                .as_ref()
                .cloned()
                .ok_or_else(|| io::Error::from(io::ErrorKind::NotFound))?
```

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:241]: this comment is unhelpful

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:356]:
```suggestion
        core::iter::from_fn(|| self.pending_updates.pop().ok())
            .take(max)
            .map(|packet| packet.update())
            .collect::<Result<Vec<()>,_>>()?;
```
the `Vec<()>` shouldn't actually allocate anything, as `()` is ZST.

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:381]: do we ever prune AFDs from `afd_handles`? could this exhaust all available memory? There currently doesn't seem to be a way to tell the poller that it should `shrink_to_fit`, free all cached blocks that aren't used anymore (or, if I understand the usage of `afd_handles` correctly, it seems to act as a cache and we could get away with
```rust
self.afd = Mutex::new(Vec::new());
```

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:410]: why isn't this `pub`?

File: src/iocp/mod.rs
Comment (fogti) [reply src/iocp/mod.rs]: [`debug_assert!`](https://doc.rust-lang.org/core/macro.debug_assert.html) gets completely eliminated on release builds. *Beware*.

File: src/iocp/mod.rs
Comment (fogti) [reply src/iocp/mod.rs:417]: hmm, always cloning the packets before the filter does seem contra-productive.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs:381]: Each AFD handle can be associated with up to 32 sockets. We use the ref-counting mechanism in `Arc` to keep track of how many sockets are being used. This ensures that old AFDs are reused by old sockets. I guess you could end up with thousands of AFD handles if you open and close 100,000's of sockets. In that case, I can add a pruning mechanism for when there are a large number of AFD handles with no associated sockets.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs:410]: `Events` is internal to the `Poller` struct and used as an event container. That being said, I'd like to make the `Events` struct public so that we don't go through a `Vec<Event>` for events, see #66.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs:417]: I think so too! This could be avoided if the `Events` struct was `pub` and could be read from directly, but this would be a breaking change. See #66.

File: src/iocp/mod.rs
Comment (fogti) [reply src/iocp/mod.rs:381]: alternatively, you could "just" use weak pointers, that way, opening, polling, and then closing 100'000's of sockets should automatically free the associated AFDs, and the next crawl of the list could then prune the gone ones.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs]: Hmm, yeah, it does look pretty ugly. I've added a `lock!` macro in order to do it automatically.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs]: I've replaced the block with a more concise version.

File: src/iocp/mod.rs
Comment (fogti) [reply src/iocp/mod.rs:440]: ok, so `Event` can actually be copied (so I think that should be used instead), then it doesn't really matter much (except that, of course, the mixing of `Events` and `Vec<Event>` instead of one just being a `type` alias of the other or such is unfortunate)

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs]: I changed it to clarify what it's doing, making it more helpful

File: src/iocp/mod.rs
Comment (fogti) [inline src/iocp/mod.rs:57]: Rust has hygienic macros, so the parentheses around `$lock_result` are unnecessary.

File: src/iocp/mod.rs
Comment (notgull) [reply src/iocp/mod.rs]: Great idea! I used `try_for_each` instead of `.map.collect` since it's more idiomatic.

Unrelated: Should we implement an `Iterator` for draining a `ConcurrentQueue` in order to avoid using `from_fn()`?

File: src/iocp/port.rs
Comment (fogti) [reply src/iocp/port.rs:88]: hm, seems a bit brittle, but I'll go with it for now.

File: src/iocp/mod.rs
Comment (fogti) [reply src/iocp/mod.rs]: would be ok, it should be documented that the `Iterator::next`, analogous to the rest of the interface, won't block.

**Comment by fogti** [comment]
Thanks for the work btw. Looks impressive.

**Review by fogti — APPROVED** [review]

**Comment by notgull** [comment]
> Thanks for the work btw. Looks impressive.

Thanks for the review! I know this is a monster PR, so I appreciate you taking the time to sit down and look through it all. I'll merge once the CI passes.

**Comment by fogti** [comment]
It might be a good idea to do another downstream test, otherwise this looks ok. I think it is appropriate to squash-merge this when done. The possible brittleness of `<Pin<Arc<T>> as CompletionHandle>::get()` (or maybe the whole `CompletionHandle` + `OVERLAPPED` API) should imo be mentioned/appended to the intro post in the resulting merge commit.
"run run" in the intro post looks like a mistake btw.

> We only expect one user to be polling at a time, so we use an AtomicBool instead of an atomic number.

Can/do we enforce that?

**Comment by notgull** [comment]
> > We only expect one user to be polling at a time, so we use an AtomicBool instead of an atomic number.
> 
> Can/do we enforce that?

A `Mutex` in the top-level `Poller` is locked before any polling is done.

https://github.com/smol-rs/polling/blob/e85331c437eb5b9fce382e6a9a310e44a641a3d6/src/lib.rs#L491-L505

**Pull request merged**
